pshtt.py

```diff
@@ -209,32 +248,155 @@ def hsts_check(endpoint):
         endpoint.hsts_preload = True
 
 
-def bad_chain(trusted, endpoint):
-    # If the cert is not trusted by mozilla it is a bad chain
-    if "FAILED" in trusted:
-        endpoint.https_bad_chain = True
-
-
-def bad_hostname(hostname_validation, endpoint):
-    # If hostname validation fails
-    if "FAILED" in hostname_validation:
-        endpoint.https_bad_hostname = True
-
+# Uses sslyze to figure out the reason the endpoint wouldn't verify.
+def https_check(endpoint):
+    logging.debug("sslyzing %s..." % endpoint.url)
 
-def expired_cert(expired_date, endpoint):
-    # Split the time into an list of subtrings
-    temp = expired_date.split()
-    # Convert the date returned by sslyze to be comparable to current time
-    if datetime.datetime(int(temp[5]), strptime(temp[2], '%b').tm_mon, int(temp[3])) < datetime.datetime.now():
-        endpoint.https_expired_cert = True
+    # remove the https:// from prefix for sslyze
+    hostname = endpoint.url[8:]
+    server_info = ServerConnectivityInfo(hostname=hostname, port=443)
+    server_info.test_connectivity_to_server()
+
+    cert_plugin = CertificateInfoPlugin()
+    cert_plugin_result = cert_plugin.process_task(server_info, 'certinfo_basic')
+
+    # A certificate can have multiple issues.
+    for msg in cert_plugin_result.as_text():
+
+        # Check for certifcate expiration.
+        if (
+            (("Mozilla NSS CA Store") in msg) and
+            (("FAILED") in msg) and
+            (("certificate has expired") in msg)
+            ):
+            endpoint.https_expired_cert = True
+
+        # Check for whether there's a valid chain to Mozilla.
+        if (
+            (("Mozilla NSS CA Store") in msg) and
+            (("FAILED") in msg) and
+            (("unable to get local issuer certificate") in msg)
+            ):
+            endpoint.https_bad_chain = True
+
+        # Check for whether the hostname validates.
+        if (
+            (("Hostname Validation") in msg) and
+            (("FAILED") in msg) and
+            (("Certificate does NOT match") in msg)
+            ):
+            endpoint.https_bad_hostname = True
```

Comment:
[I also improved the sslyze parsing logic to:

* Make the text parsing more strict in detecting each certificate error situation. Previously, the `CA Store` case would match both an invalid certifciate chain, and an expired certificate.
* Make it possible for multiple errors to be flagged. Previously, it was a set of conditionals (`if`/`elif`) that would only match the first error seen. This will now detect if a certificate has, for example, a bad chain *and* a bad name. (For context, this does happen -- we had [a bad bug in Pulse](https://github.com/18F/pulse/pull/224) where we weren't properly detecting this case, and fixing it forced us to drop the `Uses HTTPS` column by **4.5%**.